### ISSUE TITLE
Problem with xworkspaces label-urgent

### ISSUE THREAD

**Issue by jamesbvaughan** [polybar/polybar#1081]
I'm running polybar on Arch Linux with Xmonad.

The xworkspaces module correctly identifies and labels the active workspace, but labels all other workspaces as "empty". I haven't been able to get the "occupied" or "urgent" labels to show under an circumstances.

My xworkspaces config:
```dosini
[module/workspaces]
type = internal/xworkspaces
format = <label-state>
pin-workspaces = false
label-active-foreground = ${colors.foreground-alt}
label-empty-foreground = ${colors.foreground-alt}
label-active = "  "
label-empty = "  "
label-occupied = "  "
label-urgent = "  "
```

Is this a known bug? Is there a clear problem with that section of my config?

**Comment by NBonaparte** [comment]
See https://github.com/jaagr/polybar/wiki/Module:-xworkspaces, occupied states haven't been implemented yet. Urgent labels should work though, not sure why they don't work.

#882 attempts to implement occupied states, but as we have discovered, Xmonad has odd behaviour, so it might not fully work.

**Comment by jamesbvaughan** [comment]
Ahh, thanks for the link. I've looked at that page and somehow always glossed over that first section.

**Comment by sagotsky** [comment]
@NBonaparte Any suggestions for debugging the urgent label?  I'm also on arch w/ xmonad.  The closest I've gotten is to have the urgent label flash when the flag gets set and disappear immediately after.  It's the only thing keeping me on xmobar.

**Comment by NBonaparte** [comment]
You could try using `xprop -spy` to check for the `_NET_WM_STATE` property and see if it contains `_NET_WM_STATE_DEMANDS_ATTENTION` and/or `WM_HINTS` states that "The urgency hint bit is set". Also watch for when these properties change back to normal.

**Comment by sagotsky** [comment]
This is what I see when I set urgency.  I get the same results whether or not polybar is running.
```
_NET_WM_DESKTOP(CARDINAL) = 4
WM_HINTS(WM_HINTS):
_NET_WM_DESKTOP(CARDINAL) = 4
WM_HINTS(WM_HINTS):
                The urgency hint bit is set
_NET_WM_DESKTOP(CARDINAL) = 4
```

**Comment by dkasak** [comment]
@sagotsky Ever managed to solve this? It's happening to me too.

**Comment by sagotsky** [comment]
No, but it's been a while since I tried.

**Comment by dkasak** [comment]
I've tried running polybar in gdb and setting a breakpoint on `xworkspaces.cpp:113`, which is this line:

```
112│     } else if (evt->atom == WM_HINTS) {
113├─────> if (icccm_util::get_wm_urgency(m_connection, evt->window)) {
114│         set_desktop_urgent(evt->window);
115│       }
116│     } else {
```

Then I've used a [small C program](https://unix.stackexchange.com/a/320885) to make my terminal window urgent, like this:

```sh
➜ ./seturgent $(xdotool getwindowfocus)
```

This correctly triggers the breakpoint, but since the window is on the current desktop, `set_desktop_urgent` ignores it and doesn't actually set it as urgent (which is expected). However, when I do this instead

```sh
➜ window=$(xdotool getwindowfocus)
<here I switch workspaces>
➜ sleep 2; ./seturgent $window
```

the breakpoint never gets triggered. So it seems like `WM_HINTS` events for windows not on the current workspace don't get delivered to the `XCB_PROPERTY_NOTIFY` handler (`xworkspaces_module::handle`)?

I'm also on xmonad.

**Comment by NBonaparte** [comment]
Interesting. Could you test this on another WM to make sure this isn't an idiosyncrasy for xmonad?

**Comment by dkasak** [comment]
Yes. I just tried with i3 and the observed behaviour is the same. i3's own status bar correctly detects the urgent workspace by painting its icon with a red background (the default settings), but the breakpoint in polybar never gets triggered. If I run `seturgent` in a workspace while watching it, then it does get triggered (but has no effect on the xworkspaces module).

**Comment by Pascal3366** [comment]
I've been also waiting for the urgent and occupied states feature in xworkspaces / ewmh for a long time

**Comment by sagotsky** [comment]
Urgency is still being weird.  When I set a window as urgent, no matter where the window is the first workspace gets the urgent label.  This is happening on multiple WMs.

**Comment by ginsburgnm** [comment]
Hey all, here's a workaround script in python while people are waiting. 

As-is the script sets active workspace to a cyan, and urgent to a light-green.
I'm posting this as a helpful script, not a full featured module; nor as something I intend to maintain.

but, to change what colors are used look at `convert_to_string`
it sets the font color `%%{F#00acc1}` and the overline color `%%{o#00acc1}`, due to how string formatting works in python, the extra `%` is to escape it.

Anyway, hope this is useful to someone.
```
#!/usr/bin/env python3
"""Check current workspaces"""
from subprocess import run, PIPE
from json import loads

def get_workspaces():
    """shell out to i3"""
    command = ['i3-msg', '-t', 'get_workspaces']
    stdout = run(
        command, stdin=None, input=None, stdout=PIPE, stderr=None,
        shell=False, cwd=None, timeout=None, check=False, encoding=None,
        errors=None, env=None
    ).stdout
    return stdout

def ascii_to_json(ascii_value):
    """Convert ascii to json"""
    return loads(ascii_value.decode("UTF-8"))

def convert_to_string(workspaces):
    """convert to user defined string"""
    for workspace in workspaces:
        if workspace['focused']:
            workspace['name'] = "%%{F#00acc1}%%{o#00acc1}%s%%{F-}" % workspace['name']
        elif workspace['urgent']:
            workspace['name'] = "%%{o#7cb342}%%{F#7cb342}%s%%{F-}" % workspace['name']
        else:
            workspace['name'] = "%%{o#263238}%s" % workspace['name']
    return " ".join([w['name'] for w in workspaces])

def main():
    """cli"""
    print(convert_to_string(
        ascii_to_json(
            get_workspaces()
        )
    ))

if __name__ == "__main__":
    main()
```

**Comment by patrick96** [comment]
@ginsburgnm The label-urgent for the i3 module should work properly, only the xworkspaces module is affected and on i3 it's almost always easier to use the i3 module.

**Comment by ginsburgnm** [comment]
@patrick96 Yeah, label-urgent works for i3, but the formatting for the module is less than ideal compared to xworkspaces for my purposes. the separation between workspaces is clunky. So I was using xworkspaces because it looks prettier, but then I wanted to have `urgent` work. So I just wrote my own.

**Comment by sagotsky** [comment]
I'm not on i3.  I wrote my own xworkspaces replacement.  It's still pretty quick and dirty.  And it's in ruby so it's not resource efficient.  But it responds to my urgency hooks, so it's an improvement.  https://github.com/sagotsky/.dotfiles/commit/4c5662a13085926d9d726dd73e74e49024268a9e

**Comment by patrick96** [comment]
@sagotsky Sorry, for some reason I thought you used i3.

**Comment by erichamers** [comment]
I'm on Qtile with polybar and also can't get the urgent hint to work correctly, it works with the qtilebar though.

**Comment by SalmanFarooqShiekh** [comment]
Has this been solved? I am on Qtile with polybar and xworkspaces module doesn't show any indication of urgent workspaces

**Comment by patrick96** [comment]
This will be in the 3.6.0 release

**Comment by Ahanaf-Ether** [comment]
hasent been as of release; 3.6.2

**Comment by patrick96** [comment]
@Ahanaf-Ether This has been implemented in 3.6. Please open a bug report if you think the xworkspaces module is misbehaving.
